include/z64collision_check.h

```diff
@@ -9,90 +9,140 @@
 // From z64.h
 struct Actor;
 
+/*
+ * Bases for all shapes of colliders
+ */
+
+typedef enum {
+    /* 0  */ COLTYPE_HIT0, // Blue blood, white hitmark
+    /* 1  */ COLTYPE_HIT1, // No blood, dust hitmark
+    /* 2  */ COLTYPE_HIT2, // Green blood, dust hitmark
+    /* 3  */ COLTYPE_HIT3, // No blood, white hitmark
+    /* 4  */ COLTYPE_HIT4, // Water burst, no hitmark
+    /* 5  */ COLTYPE_HIT5, // No blood, red hitmark
+    /* 6  */ COLTYPE_HIT6, // Green blood, white hitmark
+    /* 7  */ COLTYPE_HIT7, // Red blood, white hitmark
+    /* 8  */ COLTYPE_HIT8, // Blue blood, red hitmark
+    /* 9  */ COLTYPE_METAL,
+    /* 10 */ COLTYPE_NONE,
+    /* 11 */ COLTYPE_WOOD,
+    /* 12 */ COLTYPE_HARD,
+    /* 13 */ COLTYPE_TREE
+} ColliderType;
+
+typedef enum {
+    /* 0 */ COLSHAPE_JNTSPH,
+    /* 1 */ COLSHAPE_CYLINDER,
+    /* 2 */ COLSHAPE_TRIS,
+    /* 3 */ COLSHAPE_QUAD,
+    /* 4 */ COLSHAPE_MAX
+} ColliderShape;
+
 typedef struct {
-    /* 0x00 */ struct Actor* actor; // Attached actor
-    /* 0x04 */ struct Actor* at; // Actor attached to what it collided with as an AT collider.
-    /* 0x08 */ struct Actor* ac; // Actor attached to what it collided with as an AC collider.
-    /* 0x0C */ struct Actor* oc; // Actor attached to what it collided with as an OC collider.
-    /* 0x10 */ u8 atFlags; // Information flags for AT collisions.
-    /* 0x11 */ u8 acFlags; // Information flags for AC collisions.
-    /* 0x12 */ u8 ocFlags1; // Information flags for OC collisions.
-    /* 0x13 */ u8 ocFlags2;  // Flags related to which colliders it can OC collide with.
-    /* 0x14 */ u8 colType; // Determines hitmarks and sound effects during AC collisions.
-    /* 0x15 */ u8 shape; // JntSph, Cylinder, Tris, or Quad
+    /* 0x00 */ struct Actor* self; // Actor owning this collider.
+    /* 0x04 */ struct Actor* otherAC; // Other actor owning the AC collider that collided with this AT collider.
+    /* 0x08 */ struct Actor* otherAT; // Other actor owning the AT collider that collided with this AC collider.
+    /* 0x0C */ struct Actor* otherOC; // Other actor owning the OC collider that collided with this OC collider.
```

Comment:
This does seem like a good idea, but I think `self` would confuse people coming from languages (Python, Rust) that use `self` to refer to the object itself. Maybe, as your comment suggests, `owner` would work?

Reply:
Sure I'm open to changing it, `owner` sounds fine

Fwiw "self" is already used in some parts of zcolchk iirc, but it can be changed there too

It's not super straightforward to change though so I'd like more opinions to consolidate the name before commiting to another rename

Reply:
I guess I don't understand what's wrong with just using `actor` here.

What's far worse in this part of code is that you've inverted the meaning of the at, ac, oc variables, which makes the code harder to follow and use correctly.  

The at, ac, oc variables are meant to store the _RESULT_ of the collision check computation requested when calling SetAT/SetAC/SetOC, relative to the reference frame of the collider itself. Did the collider's attack land? `at` stores the actor that it hit. Did the attack collider get hit? `ac` stores the actor that touched it. Did it bump into an object? `oc` stores the actor it touched.

Similarly, `atFlags`, and `acFlags` are named to be associated with the `at` and `ac` variables, since they effect the result of the collision check. This is why e.g. `Collider_ResetATBase` nulls `at` and masks `atFlags` to a default value in the same function.

Lastly, inverting the name creates some misgivings about how the system works. Using your renames, `otherAT`, `otherAC`, `otherOC`, one might assume that reading e.g. `otherAT` would return an actor currently subscribed to AT collisions. This is entirely incorrect, since the lists are purged after performing collision check (which happens after actors are processed and before the next frame when the results are available to actors), and actors are free to redefine what lists to subscribe a collider to.

Reply:
What you explain matches what I know of how zcolchk works, and I also guessed the same about what the logic behind the previous names was

If I changed them, it's because I believe the names I suggest are better.
For example when you're inside an actor's code, it's not obvious, it's confusing even, what `collider.at` means. `collider.otherAC` seemed clearer: it's another (not me), and it's an AC (I attacked it)

Reply:
`collider.at` is the result of the SetAT call you made last frame; the core advantage is that you don't have to do a mental inversion use the correct variable name since all `AT` things are named `AT`, and you're not mixing AT with AC.

and again, `otherAC` doesn't return an AC; it returns an actor or NULL. ACs no longer exist by the time you read the variable.

Reply:
> and again, otherAC doesn't return an AC; it returns an actor or NULL. ACs no longer exist by the time you read the variable.

sure? does that not apply to `at`

Reply:
It does so yes. I think the difference between us is our perspective on what the variable name is actually naming. 

The idea behind the `at` name is that it's shorthand for "This is the result Actor of this collider's AT Collision Check". The thing being called `at` is the result pointer. The actor being referenced (or NULL) isn't the thing with an `AT` property or an accessible `AT` collider. Another way of putting it is that `at` is really a handle name used to get the collider's AT Collision Check result... if things were implemented in python, `ColliderCollisionResult["AT"]` better expresses the intent behind the variable name. Given this, I find it weird to change the handle to get AT collision results to one that has AC in the name.

With `otherAC` my understanding is that this name was chosen to describe a property of the actor being stored here, e.g. dereferencing `otherAC` gives you an actor with an AC enabled collider or an actor with some sort of AC, which isn't a correct interpretation. AT/AC/OC aren't properties attached to actors, they are properties attached to colliders.

Reply:
Well, for an actor to end up in `otherAC`, it needs to use colliders and have had a collider set as AC
I didn't make it `otherACOwner` because that looks bad and seems redundant

I feel like naming it like a result works less well, `atResult` sounds less natural

How I came to pick `otherAC` was like:
1) `at`/`ac`/`oc` on its own isn't descriptive enough, it's confusing whether that refers to "self" or "other", and also reading code you can't tell it's not a part of the collider struct -> use `other`
2) `atOther` can be confused between "other actor using AT" and "other actor relevant to self being AT" -> `otherAC`

```c
    if (this->colliderBody.base.atFlags & AT_HIT) {
        this->colliderBody.base.atFlags &= ~AT_HIT;
        if (this->colliderBody.base.at == &player->actor) {
```

```c
    if (this->colliderBody.base.atFlags & AT_HIT) {
        this->colliderBody.base.atFlags &= ~AT_HIT;
        if (this->colliderBody.base.otherAC == &player->actor) {
```

I really think the second one reads better

Reply:
> Well, for an actor to end up in `otherAC`, it needs to use colliders and have had a collider set as AC I didn't make it `otherACOwner` because that looks bad and seems redundant

for NULL to end up there it has to do neither

> I feel like naming it like a result works less well, `atResult` sounds less natural

Disagree; I think it's pretty natural to think of the system working as such:

* Set AT Flags
* Enable AT collision on collider
* Wait
* Check AT collision result on collider

> at/ac/oc on its own isn't descriptive enough, it's confusing whether that refers to "self" or "other"

Why would a collider collide with itself? It should be implicit that a Collider would only collide with something else.

> I really think the second one reads better

It's slightly more readable, but it breaks the consistency that existed, making it harder to use. You could easily achieve even greater clarity with code while maintaining naming consistency like this:

    if (this->colliderBody.base.atFlags & AT_HIT) {
        this->colliderBody.base.atFlags &= ~AT_HIT;
        if (this->colliderBody.base.atHitActor == &player->actor) {

Not suggesting that `at` -> `atHitActor`, but rather that it's easier to understand why `atFlags` are being checked here in this snippet, whereas with `otherAC` it's less obvious.

Reply:
So like are you open to renaming these to anything or do you want all of them to stay as they are in master actor/at/ac/oc

Reply:
I'm personally fine with the names actor/at/ac/oc staying as they are, and I would be fine with at/ac/oc changing to be more expressive to something like atResult/acResult/ocResult. The key thing I care about is that whatever name is used, at/ac/oc all line up like before so that we avoid mixing at and ac flags.

Reply:
-> #1436